File: A02/P02/f.-CribaEratostenes.cpp
#include <iostream>
#include <algorithm>
#include <iomanip>
#include <sstream>
#include <string>
#include <vector> 
#include <ctime>
#include <cmath>
using namespace std;

string formatTime(long double t) {
	ostringstream time;
	int hour, min, sec, milisec;
	
	hour = t/3600;
	t = t - (hour*3600);
	min = t/60;
	t = t - (min*60);
	sec = t;
	t = t - (sec/1);
	milisec = t*10000;

	time << setfill('0') <<
		setw(2) << hour << ':' <<
		setw(2) << min << ':' <<
		setw(2) << sec << '.' << 
		setw(4) << milisec;
	
	return time.str();
}

void criba(long long num1, long long num2, vector<bool> &primes){
	primes = vector<bool>(num2 - num1, true);
	for(long long i = 2; i <= sqrt(num2); i++) {
		if(primes.at(i-1))
		for (long long j = i; j*i <= num2; j++) {
			primes[(i*j)-num1]=0;
		}
	}
}

bool isPrime(long long num, vector<bool> primes){
	return primes.at(num);
}

int main() {
	vector<bool> primes;
	vector<long long> numbers;
	long long number, min, max;
	int nPrimes = 0;
	clock_t t;
	string time;
	
	
	while(1){
		cin >> number;
		if (number == -1) break;
		numbers.push_back(number);
	}
	if(numbers.size() > 0){
		min = *min_element(numbers.begin(), numbers.end());
		max = *max_element(numbers.begin(), numbers.end());
		criba(min, max+1, primes);
	}
	t = clock();
	for(long long i = 0; i < numbers.size(); i++){
		nPrimes += isPrime(numbers.at(i)-min, primes);
	}
	t = clock() - t;

	/*for(int i = 0; i < primes.size(); i++) 
		cout << primes.at(i) << " ";
	cout << endl;*/

	time = formatTime(((long double)t)/CLOCKS_PER_SEC);

	cout << "Total: " << nPrimes << endl;
	cout << "Tiempo: " << time << endl;

	return 0;
}

File: A02/Python/P02/f.-CribaEratostenes.py
import time
import math

def format_time(t):
    time_formatted = ""
    hour = minute = second = milisecond = int(0)

    hour = int(t//3600);
    t = t - (hour*3600)
    minute = int(t//60)
    t = t - (minute*60)
    second = int(t//1)
    t = t - second
    t = int(t*10000)
    milisecond = t
    
    time_formatted += (("0" + str(hour)) if hour < 10 else str(hour)) + ':'
    time_formatted += (("0" + str(minute)) if minute < 10 else str(minute)) + ':'
    time_formatted += (("0" + str(second)) if second < 10 else str(second)) + '.'
    time_formatted += '{:0>4}'.format(milisecond)

    return time_formatted


def criba(num1, num2):
    primes = [True for i in range(num2 - num1)]
    i = 2
    #print(primes)

    while i <= math.sqrt(num2):
        #print("if",i)
        if primes[i-1]:
            j = i
            while (j*i) < num2:
                #print(i,'x',j,'=',j*i,(i*j)-num1)
                primes[(i*j)-num1] = False;
                j += 1
        i += 1
    #print(i,math.sqrt(num2))
    i=2
    #for v in primes:
        #print(i,v)
        #i+=1

    return primes


def main():
    mi = ma = 0
    number = 0
    numbers = []
    primes = []
    n_primes = 0
    exec_time = ""
    t = 0.0
    
    while True:
        number = int(input())
        if number == -1: break
        numbers.append(number)

    if len(numbers) != 0:
        mi = min(numbers)
        ma = max(numbers)

        if mi == ma: mi = 0

        primes = criba(mi, ma+1)

    t = time.time()
    #for val in primes.values():
        #if val: n_primes += 1
    for num in numbers:
        if primes[num-mi]:
            n_primes += 1
            #print(num)
    t = time.time() - t
    exec_time = format_time(t);
    
    #print(primes)
    print("Total:", n_primes)
    print("Tiempo:", exec_time)

if __name__ == "__main__":
    main()
